components/webvr/webvr_thread.rs

```diff
@@ -0,0 +1,349 @@
+            },
+            Err(msg) => {
+                if let Some(sender) = sender {
+                    sender.send(Err(msg.into())).unwrap();
+                }
+            }
+        }
+    }
+
+    fn handle_create_compositor(&mut self, display_id: u64) {
+        let compositor = self.service.get_display(display_id).map(|d| WebVRCompositor(d.as_ptr()));
+        self.vr_compositor_chan.send(compositor).unwrap();
+    }
+
+    fn poll_events(&mut self, sender: IpcSender<bool>) {
+        let events = self.service.poll_events();
+        if !events.is_empty() {
+            let pipeline_ids: Vec<PipelineId> = self.contexts.iter().map(|c| *c).collect();
+            for event in events {
+                let event = WebVREventMsg::DisplayEvent(event);
+                self.constellation_chan.send(ConstellationMsg::WebVREvent(pipeline_ids.clone(), event)).unwrap();
+            }
+        }
+
+        // Stop polling events if the callers are not using VR
+        self.polling_events = self.contexts.len() > 0;
+        sender.send(self.polling_events).unwrap();
+    }
+
+    fn notify_event(&self, event: VRDisplayEvent) {
+        let pipeline_ids: Vec<PipelineId> = self.contexts.iter().map(|c| *c).collect();
+        let event = WebVREventMsg::DisplayEvent(event);
+        self.constellation_chan.send(ConstellationMsg::WebVREvent(pipeline_ids.clone(), event)).unwrap();
+    }
+
+    fn schedule_poll_events(&mut self) {
+        if !self.service.is_initialized() || self.polling_events {
+            return;
+        }
+        self.polling_events = true;
+        let webvr_thread = self.sender.clone();
+        let (sender, receiver) = ipc::channel().unwrap();
+
+        thread::Builder::new().name("WebVRPollEvents".into()).spawn(move || {
+            loop {
+                if webvr_thread.send(WebVRMsg::PollEvents(sender.clone())).is_err() {
+                    // WebVR Thread closed
+                    break;
+                }
+                if !receiver.recv().unwrap_or(false) {
+                    // WebVR Thread asked to unschedule this thread
+                    break;
+                }
+                thread::sleep(time::Duration::from_millis(EVENT_POLLING_INTERVAL));
+            }
+        }).expect("Thread spawning failed");
+    }
+}
+
+/// Notes about WebVRCompositorHandler implementation:
+/// Raw pointers are used instead of Arc<Mutex> as a heavy optimization for latency reasons.
+/// This also avoids "JS DDoS" attacks: like a secondary JavaScript tab degrading performance
+/// by flooding the WebVRThread with messages while the main JavaScript tab is presenting to the headset.
+/// Multithreading won't be a problem because:
+///    * Thanks to the security rules implemented in the WebVRThread, when a VRDisplay is in a presenting loop
+///      no other JSContext is granted access to the VRDisplay. So really there aren’t multithreading race conditions.
```

Comment:
I assume that this is the `access_check` method above? If so, please make a comment about its importance near that method so people don't accidentally remove / "optimize" it away :-)